### ISSUE TITLE
Fail to run HiBench and indicates load Spark master failed and no INPUT_HDFS defined

### ISSUE THREAD

**Issue by chinadragon0515** [intel-hadoop/HiBench#102]
I have mapr cluster, I set these properties in 99-user_defined_properties.conf,
# Hadoop home
hibench.hadoop.home             /opt/mapr/hadoop/hadoop-2.5.1/
# Spark home
hibench.spark.home              /opt/mapr/spark/spark-1.2.1/
# HDFS master
hibench.hdfs.master             maprfs:///

And I change the benchmarks.lst to workcount only and languages.lst to mapreduce only, I meet two issues, any other configuration I need to do for this testing?

./bin/run-all.sh 
Prepare wordcount ...
Exec script: /mnt/mapr/HiBench-master/workloads/wordcount/prepare/prepare.sh
Parsing conf: /mnt/mapr/HiBench-master/conf/00-default-properties.conf
Parsing conf: /mnt/mapr/HiBench-master/conf/10-data-scale-profile.conf
Parsing conf: /mnt/mapr/HiBench-master/conf/99-user_defined_properties.conf
Parsing conf: /mnt/mapr/HiBench-master/workloads/wordcount/conf/00-wordcount-default.conf
Parsing conf: /mnt/mapr/HiBench-master/workloads/wordcount/conf/10-wordcount-userdefine.conf
Traceback (most recent call last):
  File "/mnt/mapr/HiBench-master/bin/functions/load-config.py", line 440, in <module>
    load_config(conf_root, workload_root, workload_folder)
  File "/mnt/mapr/HiBench-master/bin/functions/load-config.py", line 154, in load_config
    generate_optional_value()
  File "/mnt/mapr/HiBench-master/bin/functions/load-config.py", line 358, in generate_optional_value
    assert 0, "Get workers from spark master's web UI page failed, reason:%s\nplease set `hibench.masters.hostnames` and `hibench.slaves.hostnames` manually" % e
AssertionError: Get workers from spark master's web UI page failed, reason:Unknown resourcemanager, please check `hibench.hadoop.configure.dir` and "yarn-site.xml" file
please set `hibench.masters.hostnames` and `hibench.slaves.hostnames` manually
/mnt/mapr/HiBench-master/bin/functions/workload-functions.sh: line 33: .: filename argument required
.: usage: . filename [arguments]
start HadoopPrepareWordcount bench
/mnt/mapr/HiBench-master/workloads/wordcount/prepare/prepare.sh: line 25: INPUT_HDFS: unbound variable
ERROR: wordcount prepare failed!
Run all done!

**Comment by lvsoft** [comment]
Hibench will probe master hostname and slaves hostnames by accessing YARN's conf(`YARN_DIR/etc/hadoop/yarn-site.xml`) and YARN-CLIand(`yarn node -list`) in YARN mode. It seems hibench failed to parse the resourcemanger address from your `yarn-site.xml`. Maybe a bug, can you attach you `yarn-site.xml` for further analyze?

And meanwhile, you can define `hibench.masters.hostnames   hostname1 hostname2 hostname3 ...` and `hibench.slaves.hostnames   hostname1 hostname2 hostname3 ...` to bypass the probe of  master & workers list.

**Comment by chinadragon0515** [comment]
1. I add these two lines in 99-user_defined_properties.conf
hibench.masters.hostnames h07 
hibench.slaves.hostnames h06 h07 h08 h09 h10
I got the same error, not sure what's wrong here?
Prepare wordcount ...
Exec script: /mnt/mapr/HiBench-master/workloads/wordcount/prepare/prepare.sh
Parsing conf: /mnt/mapr/HiBench-master/conf/00-default-properties.conf
Parsing conf: /mnt/mapr/HiBench-master/conf/10-data-scale-profile.conf
Parsing conf: /mnt/mapr/HiBench-master/conf/99-user_defined_properties.conf
Parsing conf: /mnt/mapr/HiBench-master/workloads/wordcount/conf/00-wordcount-default.conf
Parsing conf: /mnt/mapr/HiBench-master/workloads/wordcount/conf/10-wordcount-userdefine.conf
Traceback (most recent call last):
  File "/mnt/mapr/HiBench-master/bin/functions/load-config.py", line 440, in <module>
    load_config(conf_root, workload_root, workload_folder)
  File "/mnt/mapr/HiBench-master/bin/functions/load-config.py", line 154, in load_config
    generate_optional_value()
  File "/mnt/mapr/HiBench-master/bin/functions/load-config.py", line 358, in generate_optional_value
    assert 0, "Get workers from spark master's web UI page failed, reason:%s\nplease set `hibench.masters.hostnames` and `hibench.slaves.hostnames` manually" % e
AssertionError: Get workers from spark master's web UI page failed, reason:Unknown resourcemanager, please check `hibench.hadoop.configure.dir` and "yarn-site.xml" file
please set `hibench.masters.hostnames` and `hibench.slaves.hostnames` manually
/mnt/mapr/HiBench-master/bin/functions/workload-functions.sh: line 33: .: filename argument required
.: usage: . filename [arguments]
start HadoopPrepareWordcount bench
/mnt/mapr/HiBench-master/workloads/wordcount/prepare/prepare.sh: line 25: INPUT_HDFS: unbound variable
ERROR: wordcount prepare failed!
Run all done!

2. This is command yarn node -list result
yarn node -list
15/06/09 08:15:56 INFO client.MapRZKBasedRMFailoverProxyProvider: Updated RM address to h07.test.com/10.249.103.77:8032
Total Nodes:6
         Node-Id	     Node-State	Node-Http-Address	Number-of-Running-Containers
h10.test.com:57894	        RUNNING	h10.test.com:8042	                           0
h07.test.com:59294	        RUNNING	h07.test.com:8042	                           0
h05.test.com:53050	        RUNNING	h05.test.com:8042	                           0
h06.test.com:42655	        RUNNING	h06.test.com:8042	                           0
h08.test.com:53695	        RUNNING	h08.test.com:8042	                           0
h09.test.com:37661	        RUNNING	h09.test.com:8042	                           0

3. Here is the content of yarn-site.xml, it does not have needed properties.
cat ./hadoop/hadoop-2.5.1/etc/hadoop/yarn-site.xml
<?xml version="1.0"?>

<configuration>
  
  <property>
    <name>yarn.resourcemanager.ha.custom-ha-enabled</name>
    <value>true</value>
    <description>MapR Zookeeper based RM Reconnect Enabled. If this is true, set the failover proxy to be the class MapRZKBasedRMFailoverProxyProvider</description>
  </property>
  <property>
    <name>yarn.client.failover-proxy-provider</name>
    <value>org.apache.hadoop.yarn.client.MapRZKBasedRMFailoverProxyProvider</value>
    <description>Zookeeper based reconnect proxy provider. Should be set if and only if mapr-ha-enabled property is true.</description>
  </property>
  <property>
    <name>yarn.resourcemanager.recovery.enabled</name>
    <value>true</value>
    <description>RM Recovery Enabled</description>
  </property>

  
</configuration>

4. I configure spark to run in yarn-client mode, do we still need spark master and slaves? Should we skip spark master and slave parse if we run spark in yarn client mode?

5.  I see load-config.py will parse xml for property yarn.resourcemanager.address, how we plan to support for RM configured with HA case? There will be property yarn.resourcemanager.address.{id}

6. Is INPUT_HDFS not related to yarn-site.xml?

**Comment by chinadragon0515** [comment]
1. I did one more try, I change the load-config.py to hard code master in that file, I can run the program.But meet two issues,
I. Seems it try to access other systems which requires password input (I use non root user to run), can we avoid this?
II. Some permission denied
/mnt/mapr/HiBench-master/bin/functions/workload-functions.sh: line 110: /dev/stderr: Permission denied
Deleted maprfs:///HiBench/Wordcount/Output
/mnt/mapr/HiBench-master/bin/functions/workload-functions.sh: line 122: /dev/stderr: Permission denied

2. How we control the test data sizing like data sizing for workcount? I see by default we have about 30GB data.

3. Do we have any reference performance data on the performance like what's the expected performance for wordcount with x nodes and nodes configuration are xxx

**Comment by lvsoft** [comment]
Oh, I rechecked the code according to your response, and found a bug which failed to skip master/slaves hostname probe. I'll create an issue to track and fix this. Many thanks!

1. During running, hibench will monitor system utilization of your worker nodes. That's why it'll ssh to every your worker nodes and execute a small script remotely. This behavior assume you have installed your ssh public key to `~/.ssh/authorized_keys` in each node (including localhost) . And this is also a mandatory step required by hadoop & spark distribution installation. So I suggest you install your ssh public key to avoid password prompt.

  `/dev/stderr: Permission denied` should be a configuration issue of your system. Every user is assumed to  have write permission to `/dev/stderr` under common circumstances.

2. You can edit/add in `conf/10-data-scale-profile.conf` to adjust/add data sizes of each workload. And edit `conf/99-user_defined_properties.conf` for the value of `hibench.scale.profile` to choose different workload size scale profile defined in `conf/10-data-scale-profile.conf`

3. Yes, you can refer to http://intel-bigdata.github.io/ASP/result.html which is currently under working and will release benchmark result against spark upstream every day. However, currently we only reveal normalized results. I can give you a detailed result personally if you interested. Still, please note that performance will vary greatly according to different system configurations and workload parameters.

**Comment by chinadragon0515** [comment]
Lv Qi, thanks very much for your help.
1. I have passwordless access for root, but no passwordless access for user hdfs, and I need to run yarn job with user hdfs. Then I met the issues.
I modify the code, I run ./run-all.sh with root account,  but in the yarn submit, I use sudo -u hdfs to submit the job, then I did not see any issues.

Then question are do we required root access to run or support non-root account to run?

2.& 3, thanks.
